### PR TITLE
Fixed: Array values are wrapped with double quotes

### PR DESCRIPTION
**Pull request by MuratUrsavas**
## What does this PR do?
Fixes wrapping array values with extra double quotes.

## GitHub issue or Phabricator ticket number?
https://phorge.wireload.net/T7534

## Checklist before merging

- [x] If have added tests.
- [ ] Is this a new feature? If yes, please write one phrase about this update.
- [ ] I've attached relevant screenshots (if relevant).

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "cli"
-version = "0.2.0"
+version = "0.2.1"
```
Comment (sergey-borovkov) [inline Cargo.toml:3]: Why ?

File: src/cli.rs [hunk 2]
```diff
@@ -1148,9 +1148,13 @@ pub fn handle_cli_edge_app_command(command: &EdgeAppCommands) {
             generate_mock_data,
         } => {
             let secrets = if let Some(secret_pairs) = secrets {
-                secret_pairs.secrets.clone()
+                secret_pairs
```
Comment (sergey-borovkov) [inline src/cli.rs:1151]: why does it need to be converted here? And I don't see any reason why secret pairs need to be of this type. Especially in exposed API.

Any specific reason?

File: src/commands/edge_app_server.rs [hunk 3]
```diff
@@ -96,39 +102,75 @@ impl Default for Metadata {
...
+
+#[derive(Debug, Default, Deserialize, Clone)]
+struct MockDataStr {
     metadata: Metadata,
     settings: HashMap<String, String>,
 }
 
+impl MockData {
+    fn new_from_str(mock_data_str: &MockDataStr) -> Self {
+        let settings_val = mock_data_str
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:117]: why are we not using serde to deserialize string?

**Review by sergey-borovkov — CHANGES_REQUESTED** [review]

**Comment by sergey-borovkov** [comment]
Also there is no new test that tests for issue you fixed.

File: Cargo.toml (hunk 1 above)
Comment (MuratUrsavas) [reply Cargo.toml:3]: Because the code has changed. We're using versions to tag a specific state of software. If that state changes, the version has to change.

Comment (sergey-borovkov) [reply Cargo.toml:3]: No. We bump version when we release new version, not for every fix (this is not even enough as the version is referenced in docker and somewhere else)

File: src/cli.rs (hunk 2 above)
Comment (MuratUrsavas) [reply src/cli.rs:1151]: No specific reason actually. Was just quickly following the reference chain. Maybe overdone it :smile:

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:117]: Well, that was the failing part. It cannot automatically create the new `MockData` from a string. I needed to create that middle step to fix it. If you have a more idiomatic idea, I'm all ears.

**Comment by MuratUrsavas** [comment]
> Also there is no new test that tests for issue you fixed.

Current tests already cover it. Maybe need to modify the current mock data a little bit.

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:117]: What's the issue? It's not enought data for me that it's a failing part. But if some field fails to parse correctly you can always use custom deserializer - take a look at EdgeAppManifest for instance it does that a lot

**Comment by sergey-borovkov** [comment]
> > Also there is no new test that tests for issue you fixed.
> 
> Current tests already cover it. Maybe need to modify the current mock data a little bit.

We still should cover the exact failure we had. Those tests did not reveal the issue. I'd propose to have a test that checks that the whole output is correct by comparing with the whole output string typed out in the test

**Comment by MuratUrsavas** [comment]
> We still should cover the exact failure we had. Those tests did not reveal the issue. I'd propose to have a test that checks that the whole output is correct by comparing with the whole output string typed out in the test

Will do that with the mock data change.

File: src/cli.rs (hunk 2 above)
Comment (sergey-borovkov) [reply src/cli.rs:1151]: I feel like it's a bit more natural for us not to expose those internal types they can be private to the module so that consumers don't have to care about it

File: Cargo.toml (hunk 1 above)
Comment (MuratUrsavas) [reply Cargo.toml:3]: How about using dash notation for intermediary versions under development? Like `0.2.1-1` and while releasing we can get rid of the dash suffix.

File: src/cli.rs (hunk 2 above)
Comment (MuratUrsavas) [reply src/cli.rs:1151]: Actually, the exposed part is still the same. Are you sure we're exposing the new type? Maybe I'm overseeing something?

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:117]: I'd like to learn that custom deserializer. But we need to ship this one quickly. How about implementing it as a follow-up?

File: Cargo.toml (hunk 1 above)
Comment (sergey-borovkov) [reply Cargo.toml:3]: I don't see any benefit to that as you can't download it from any place. And it's what we do in all our projects so what's the point of doing differently here?

File: src/cli.rs (hunk 2 above)
Comment (sergey-borovkov) [reply src/cli.rs:1151]: EdgeAppServerValue::Str is exposed which is a private enum

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:117]: ok since we need it fast - please create a GH issue and reference it there

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "cli"
-version = "0.2.0"
+version = "0.2.1-1"
```
Comment (sergey-borovkov) [inline Cargo.toml:3]: pls no :)

File: Cargo.toml (hunk 1 above)
Comment (MuratUrsavas) [reply Cargo.toml]: We're testing it here and there and we should be providing it to other teams for tests before releasing it. And every software iteration has to carry a different tag. Reports have to include that version info to match the exact state of that software. You can use build number, commit hash, or anything else. But some info has to indicate that state. Version is the easiest one.

Comment (sergey-borovkov) [reply Cargo.toml]: When it's unreleased that's what you have git commit hash for. We are *not* providing that as a binary for testing to anyone.

File: src/commands/edge_app_server.rs [hunk 5]
```diff
@@ -227,16 +285,21 @@ settings:
         assert!(content.contains("metadata: {")); // Make sure metadata is in its own section
         assert!(content.contains("\"srly-t6kb0ta1jrd9o0w\"")); // Checking an example metadata value
         assert!(content.contains("key"));
+        assert!(content.contains("[\"37.3861\", \"-122.0839\"]"));  // Make sure arrays are written correctly
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:288]: Let's test the whole file content, please. 

instead of contains using strict eq. 

I did this approach and it was an error since it turned out to cause bugs let's not use it further.

File: src/cli.rs (hunk 2 above)
Comment (MuratUrsavas) [reply src/cli.rs]: Done!

File: Cargo.toml (hunk 4 above)
Comment (MuratUrsavas) [reply Cargo.toml:3]: OK :)

File: src/commands/edge_app_server.rs (hunk 5 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:290]: OK, let's do this one as a follow-up. I've done it but it kept invalidating the valid string generated by itself. Will take a deeper look after that. Approve and release as it is.

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:117]: Will do.

File: src/commands/edge_app_server.rs [hunk 6]
```diff
@@ -214,29 +269,39 @@ settings:
...
 
+        print!("{}", content);
+
         assert!(content.contains("screenly ="));
         assert!(content.contains("http://127.0.0.1:8080")); // Assuming this is expected from your server
         assert!(content.contains("metadata: {")); // Make sure metadata is in its own section
         assert!(content.contains("\"srly-t6kb0ta1jrd9o0w\"")); // Checking an example metadata value
         assert!(content.contains("key"));
+        assert!(content.contains("[\"37.3861\", \"-122.0839\"]"));  // Make sure arrays are written correctly
+        assert!(!content.contains("\"[\"37.3861\", \"-122.0839\"]\""));  // Make sure arrays are not written incorrectly
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:291]: don't check for negatives, please...

File: src/commands/edge_app_server.rs (hunk 5 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:290]: I can't approve when the test for the actual content does not pass :( Maybe someone else will

File: src/commands/edge_app_server.rs [hunk 7]
```diff
@@ -138,40 +180,53 @@ fn format_js(data: MockData, secrets: &HashMap<String, String>) -> String {
...
-            "[{}]",
+        Value::Array(
             metadata
                 .tags
                 .iter()
-                .map(|tag| format!("\"{}\"", tag))
-                .collect::<Vec<String>>()
-                .join(", ")
+                .map(|tag| format!("{}", tag))
+                .collect::<Vec<String>>(),
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:208]: Clippy warning needs to be fixed

File: src/commands/edge_app_server.rs [hunk 8]
```diff
@@ -96,39 +102,75 @@ impl Default for Metadata {
...
+            .iter()
+            .map(|(k, v)| {
+                if v.len() > 2
+                    && v.chars().nth(0) == Some('[')
+                    && v.chars().nth(v.len() - 1) == Some(']')
+                {
+                    let v = &v[1..(v.len() - 1)];
+                    let v = v
+                        .split(',')
+                        .map(|s| s.trim().replace('"', "").replace('\'', ""))
```
Comment (renatgalimov) [inline src/commands/edge_app_server.rs:128]:
```suggestion
                        .map(|s| s.trim().replace(['"', '\''], ""))
```

File: src/commands/edge_app_server.rs [hunk 9]
```diff
@@ -138,40 +180,53 @@ fn format_js(data: MockData, secrets: &HashMap<String, String>) -> String {
...
-        format!(
-            "[{}]",
+        Value::Array(
             metadata
                 .tags
                 .iter()
-                .map(|tag| format!("\"{}\"", tag))
-                .collect::<Vec<String>>()
-                .join(", ")
+                .map(|tag| format!("{}", tag))
```
Comment (renatgalimov) [inline src/commands/edge_app_server.rs:207]:
```suggestion
                .map(|tag| format!(tag.to_string()))
```

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:117]: But there is no created issue and it's not referenced here :( So it's not resolved

File: src/commands/edge_app.rs [hunk 10]
```diff
@@ -286,15 +288,20 @@ impl EdgeAppCommand {
...
+    ) -> Result<(), anyhow::Error> {
         let address_shared = Arc::new(Mutex::new(None));
         let address_clone = address_shared.clone();
 
         let runtime = tokio::runtime::Runtime::new().unwrap();
         let path = path.to_path_buf();
         runtime.block_on(async {
             tokio::spawn(async move {
-                let address = run_server(path.as_path(), secrets).await.unwrap();
+                let secrets_val = secrets.iter().map(|(k, v)| { (k.clone(), EdgeAppServerValue::Str(v.clone()))}).collect();
```
Comment (korvyashka) [inline src/commands/edge_app.rs:303]: Shouldnt fmt format this line?

File: src/commands/edge_app_server.rs (hunk 5 above)
Comment (renatgalimov) [reply src/commands/edge_app_server.rs:290]: I agree with @sergey-borovkov, just do one equality test to a full file content.

File: src/commands/edge_app_server.rs [hunk 11]
```diff
@@ -13,9 +13,15 @@ use warp::{Filter, Rejection, Reply};
...
+#[derive(Debug, Clone, Deserialize)]
+pub enum Value {
+    Str(String),
+    Array(Vec<String>),
+}
+
 pub async fn run_server(
     path: &Path,
-    secrets: Vec<(String, String)>,
+    secrets: Vec<(String, Value)>,
```
Comment (korvyashka) [inline src/commands/edge_app_server.rs:24]: Hm - inst this strictly cli related format, may be pass it with string to parse later?

File: src/commands/edge_app_server.rs (hunk 9 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:207]: you don't need format! then though?

**Review by korvyashka — COMMENTED** [review]
Commented from mobile, so could miss or missunderstood smth.

1) please always do fmt and fix all clippy warnings
2) All tests on pr should pass before merging - not sure if they are, just saw the concersation

File: src/commands/edge_app_server.rs (hunk 11 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:24]: yeah, I  also think it should all be in the server code.

File: src/commands/edge_app_server.rs (hunk 9 above)
Comment (renatgalimov) [reply src/commands/edge_app_server.rs:207]: You are absolutely right

File: src/commands/edge_app.rs (hunk 10 above)
Comment (MuratUrsavas) [reply src/commands/edge_app.rs:303]: Yeah forgot to run rustfmt while on my way out. Done.

File: src/commands/edge_app_server.rs [hunk 12]
```diff
@@ -96,39 +102,75 @@ impl Default for Metadata {
...
+}
+
+#[derive(Debug, Default, Deserialize, Clone)]
+struct MockDataStr {
     metadata: Metadata,
     settings: HashMap<String, String>,
 }
 
+impl MockData {
+    fn new_from_str(mock_data_str: &MockDataStr) -> Self {
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:116]: this function also needs to be tested especially considering that you parse it manually. You need to test it with invalid input/edge cases.

File: src/commands/edge_app_server.rs (hunk 11 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs (hunk 9 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs (hunk 8 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs (hunk 5 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: OK, how a test could fail even with its own output? Would you like me to wait to figure that bug out? I'm copying its left output to the right, and then assert it.

File: src/commands/edge_app_server.rs (hunk 7 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:213]: Done!

File: src/commands/edge_app_server.rs (hunk 3 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:122]: https://github.com/Screenly/cli/issues/119 :smile:

File: src/commands/edge_app_server.rs (hunk 5 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done.

File: src/commands/edge_app_server.rs [hunk 13]
```diff
@@ -34,16 +45,16 @@ pub async fn run_server(
             },
         );
 
-    let secrets_map: HashMap<_, _> = secrets.into_iter().collect();
+    let secrets_map: BTreeMap<_, _> = secrets_val.into_iter().collect();
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:48]: Hmm. Maybe better to just use Vec<>? It preserves the order and we don't access the values by key/value

File: src/commands/edge_app_server.rs [hunk 14]
```diff
@@ -221,12 +288,8 @@ settings:
             .await
             .unwrap();
         let content = resp.text().await.unwrap();
-
-        assert!(content.contains("screenly ="));
-        assert!(content.contains("http://127.0.0.1:8080")); // Assuming this is expected from your server
-        assert!(content.contains("metadata: {")); // Make sure metadata is in its own section
-        assert!(content.contains("\"srly-t6kb0ta1jrd9o0w\"")); // Checking an example metadata value
-        assert!(content.contains("key"));
+        let expected_content = "var screenly = {\n    metadata: {\n        \"coordinates\": [\"37.3861\", \"-122.0839\"],\n        \"hardware\": \"x86\",\n        \"hostname\": \"srly-t6kb0ta1jrd9o0w\",\n        \"location\": \"Silicon Valley, USA\",\n        \"screen_name\": \"Code Cafe Display\",\n        \"tags\": [\"All Screens\"]\n    },\n    settings: {\n        \"enable_analytics\": \"true\",\n        \"override_timezone\": \"\",\n        \"tag_manager_id\": \"\"\n    },\n    secrets: {\n        \"key\": \"value\"\n    },\n    cors_proxy_url: \"http://127.0.0.1:8080\"\n};";
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:291]: maybe format it as multiline string to make it more readable?

File: src/commands/edge_app_server.rs [hunk 15]
```diff
@@ -259,4 +322,21 @@ settings:
 
         assert_eq!(resp.status(), 404);
     }
+
+    #[tokio::test]
+    async fn test_mock_data_from_str() {
```
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:327]: Please improve the naming - see here how we name our tests https://phorge.wireload.net/w/developer_guidelines/tests/

should contain what it tests and the expected behavior (but see the document anyway for in depth explanation and examples)

File: src/commands/edge_app_server.rs (hunk 13 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:48]: I think that could be left as it is.

File: src/commands/edge_app_server.rs (hunk 14 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:291]: Done!

File: src/commands/edge_app_server.rs (hunk 15 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs (hunk 13 above)
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:48]: We are not using it as a map but OK

**Review by sergey-borovkov — APPROVED** [review]

File: src/commands/edge_app_server.rs (hunk 13 above)
Comment (StanIwan) [reply src/commands/edge_app_server.rs:48]: I'm with Sergey on this one, even though BTreeMap might work, if we can also just use Vec, that would be preferrable.

**Review by StanIwan — APPROVED** [review]
I approve, this are clippy warnings I get on 1.70:
```
siwan@siwan-x1-carbon:~/Projects/cli$ cargo clippy
warning: redundant clone
   --> src/commands/edge_app_server.rs:137:47
    |
137 |                     (k.clone(), Value::Array(v.clone()))
    |                                               ^^^^^^^^ help: remove this
    |
note: this value is dropped without further use
   --> src/commands/edge_app_server.rs:137:46
    |
137 |                     (k.clone(), Value::Array(v.clone()))
    |                                              ^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
    = note: `#[warn(clippy::redundant_clone)]` on by default

warning: called `.nth(0)` on a `std::iter::Iterator`, when `.next()` is equivalent
   --> src/commands/edge_app_server.rs:129:24
    |
129 |                     && v.chars().nth(0) == Some('[')
    |                        ^^^^^^^^^^^^^^^^ help: try calling `.next()` instead of `.nth(0)`: `v.chars().next()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_nth_zero
    = note: `#[warn(clippy::iter_nth_zero)]` on by default

warning: `cli` (bin "screenly") generated 2 warnings (run `cargo clippy --fix --bin "screenly"` to apply 2 suggestions)
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
```

File: src/commands/edge_app_server.rs (hunk 13 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:48]: Yeah, I know but there is a great possibility we could use it soon. Also, it doesn't create any dev overhead. That's why I opted to leave as it is.

Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:48]: @MuratUrsavas what possibility? I am not aware of it

File: Cargo.toml (hunk 1 above)
Comment (renatgalimov) [reply Cargo.toml]: Guys. We use semantic versioning. We modify the version when we release. 
If you for whatever reason want to modify the version between releases you can use -rc tail.

It mains release candidate. 

2.0.0-rc1. 
2.0.0-rc2.
2.0.0-rc3.

We also can release the same version multiple times. We had to do it recently, to trigger the upgrade on UC16 devices. Then you can add -r tail

It means - revision.

2.0.0-r1 means 2.0.0 revision 1.

File: src/commands/edge_app_server.rs (hunk 13 above)
Comment (korvyashka) [reply src/commands/edge_app_server.rs:48]: BTree is usually used in the databases(indexes) for fast reads(searches) specifically, there is usually no reason using it in the day to day code, unless there are a lot of items in the memory you want to search in fast. So I don't see any pros using it here.

**Review by renatgalimov — APPROVED** [review]

File: src/commands/edge_app_server.rs (hunk 13 above)
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: OK, guys, you won. I've converted most of the BTreeMaps to Vec's but kept in YAML conversion. Without a custom deserializer, I cannot change it to a Vec. It will be done in another task.

**Comment by vpetersson** [comment]
Woop woop!
